### PR TITLE
Add optional arg `download_path` to the Model's constructor. Disable model download overriding.

### PR DESCRIPTION
**Pull request by dbogunowicz**
### Making sure that the changes do not break our main repos

- [x] Testing with integrations (transformers, yolo, yolact)
- [x] Testing with SparseML
- [x] Testing with Deepsparse

### Manual testing

```python
from sparsezoo import Model

stubs_to_test = [

    "zoo:cv/classification/mobilenet_v1-1.0/pytorch/sparseml/imagenet/pruned_quant-moderate",
    "zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/pruned95_obs_quant-none",
    "zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned_quant-aggressive_94"
]

for stub in stubs_to_test:
    print(stub)
    model = Model(stub, download_path = '/home/damian/testing' + stub.split("/")[-1])
    print(f'Model (from stub) path: {model.path}')
    assert model.validate(validate_onnxruntime = False, minimal_validation=True)

    model_ = Model(model.path)
    print(f'Model (from dir) path: {model_.path}')
    assert model_.validate(validate_onnxruntime=False, minimal_validation=True)
 ```

image

### CODE REVIEW

File: notebooks/model_download.ipynb [hunk 1]
```diff
@@ -88,7 +88,7 @@
     "## Selecting a model from a retrieved stub ##\n",
     "\n",
     "stub = \"zoo:cv/classification/vgg-19/pytorch/sparseml/imagenet/pruned-moderate\"\n",
-    "model = Model(stub)\n",
+    "model = Model(stub, download_path = download_path)\n",
```
Comment (bfineran) [inline notebooks/model_download.ipynb:91]:
```suggestion
    "model = Model(stub, download_path=download_path)\n",
```

File: src/sparsezoo/model/model.py [hunk 2]
```diff
@@ -68,17 +68,25 @@ class Model(Directory):
...
+            if download_path is not None:
+                path = download_path
         else:
             # initializing the model from the path
             files, path, url = self.initialize_model_from_directory(self.source)
+            if download_path is not None:
+                raise ValueError(
+                    "Ambiguous input to the constructor. "
+                    "When attempting to create Model from a local directory path, "
+                    f"`download_path` argument should be None, not {download_path}!"
```
Comment (bfineran) [inline src/sparsezoo/model/model.py:88]:
```suggestion
                    f"`download_path` argument should be None, not {download_path}"
```

File: src/sparsezoo/model/model.py [hunk 3]
```diff
@@ -68,17 +68,25 @@ class Model(Directory):
...
+    def __init__(self, source: str, download_path: Optional[str] = None):
 
         self.source = source
         self._stub_params = {}
 
         if self.source.startswith(ZOO_STUB_PREFIX):
             # initializing the files and params from the stub
             files, path, url = self.initialize_model_from_stub(self.source)
+            if download_path is not None:
+                path = download_path
```
Comment (bfineran) [inline src/sparsezoo/model/model.py:80]:
```suggestion
                path = download_path  # overwrite cache path with user input
```

File: src/sparsezoo/model/model.py [hunk 4]
```diff
@@ -223,43 +231,30 @@ def generate_outputs(
...
-            and not override
-        ):
+        # downloading is not possible
+        if os.path.isdir(download_path) and os.listdir(download_path):
             raise ValueError(
-                "Model class object was either created "
-                "using path that points to a local directory or "
-                "`download()` method already invoked."
-                "Set `override` = True to override."
+                "Attempting to download the model files to already existing directory!"
```
Comment (bfineran) [inline src/sparsezoo/model/model.py:251]:
```suggestion
                f"Attempting to download the model files to already existing directory {download_path}"
```

**Review by bfineran — DISMISSED** [review]

File: tests/sparsezoo/model/test_utils.py [hunk 5]
```diff
@@ -146,7 +146,7 @@ def test_setup_model_from_paths(self, setup):
 
     def test_setup_model_from_objects(self, setup):
         model, temp_dir, download_dir = setup
-        model.download(download_dir.name)
+        model.path = download_dir.name
```
Comment (bfineran) [inline tests/sparsezoo/model/test_utils.py:149]: ?

File: src/sparsezoo/model/model.py [hunk 6]
```diff
@@ -223,43 +237,35 @@ def generate_outputs(
...
-        directory_path: Optional[str] = None,
-        override: bool = False,
         strict_mode: bool = False,
     ) -> bool:
         """
         Attempt to download the files given the `url` attribute
-        of the files inside the Model.
+        of the files inside the Model. To avoid the risk of corrupting
+        the files while downloading, they are firstly downloaded to
+        a temporary directory, validated and finally copied over to the
```
Comment (dbogunowicz) [inline src/sparsezoo/model/model.py:246]:
```suggestion
        a temporary directory, validated, and finally copied over to the
```

**Review by markurtz — DISMISSED** [review]

**Review by KSGulin — DISMISSED** [review]

**Comment by dbogunowicz** [comment]
@markurtz  @bfineran 
After giving it more thought, I have decided to put the additional feature-saving model into temp dir on download and then validate before copying over to the target dir - on hold. To avoid overcomplicating and messy code,  this functionality has to be implemented on the base object level (inside File and Directory) rather than only in the `download()` method of the Model. I need to give it more thought, plus it is a change big enough to deserve its own PR and more testing. Let's merge this PR without the additional feature.

**Review by KSGulin — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**